[ADR]Introduce Specification Pattern (#13812)
Sylius/Sylius: adr/2022_03_31_specification_pattern_introduced.md

# Introduction of Specification Pattern

* Status: proposed
* Date: 2022-03-31

## Context and Problem Statement

> During the implementation of Headless Checkout [Pull Request](https://github.com/Sylius/Sylius/pull/13793) we encountered
> need to use a Specification pattern as multiple business requirements must be met before completing checkout.

Reviewer: ```suggestion
During the implementation of a Headless Checkout ([Pull Request](https://github.com/Sylius/Sylius/pull/13793)) we encountered the need to enforce multiple business requirements when completing the checkout, which is a perfect application of the [Specification pattern](https://en.wikipedia.org/wiki/Specification_pattern).
```

> ### What Headless Checkout is?

Reviewer: ```suggestion
### What is Headless Checkout?
```

> The current checkout system is highly coupled with UI, therefore to maintain the current checkout system, but use it via API
> a user had to make at least 5 requests in a specific order (add to cart, address, payment select, shipping select, complete).
> The new concept would reduce the number of requests to the API and get rid of the order requirement.

Reviewer: ```suggestion
The current checkout system is highly coupled with UI. To use it via API, multiple requests must be made in a specific order: add to cart, address, select payment, select shipping, complete.
The new concept would allow completing checkout with fewer requests in any order.
This is crucial for one-page checkouts, where the user might perform any action in any order.
```

## Decision Drivers

> * User must be able to create, change and extend the logic of completing order requirements

Reviewer: ```suggestion
* User must be able to add, remove, or change order requirements
```

> * Solution may be generic (abstract) from implementation

Reviewer: ```suggestion
* The solution may consist of a generic (abstract) and a concrete implementation
```

> * Specification or Composite must be able to be passed via Dependency Injection

Reviewer: ```suggestion
* The solution must be applied using Dependency Injection to facilitate its customization by the user
```
